elastic/kibana
# [App Search] Implement various Relevance Tuning states and form actions

## Summary

The commits in this PR are atomic, you should be able to follow along commit by commit.

1. **Added Save and Reset buttons**

    To test:
    - The Reset button should clear all current boosts and weights
    - The Save button should save your current relevance changes and show a success message
    - These buttons should *not* appear if you have not yet uploaded documents to an engine
   
2. **Added the `UnsavedChangesPrompt` to prevent users from leaving the form if they have unsaved changes** 

    To test:
    - Try navigating away from the page when there are changes

3. **Re-arranged the header to use the new `description`, `body`, and `action` fields**
4. **Added an Invalid boosts error message**

    To test:
    - Change a field to "number" (in old UI "Schema" screen) and create a functional boost
    - Switch the schema back to "text" (in old UI "Schema" scree
    - In the new UI, you should see the following message
    - In the new UI, delete this boost and click save, the message should go away

<img width="1707" alt="invalid_boosts" src="https://user-images.githubusercontent.com/1427475/109025397-b212af00-768c-11eb-94ca-644358b902a5.png">

5. **Added an unconfirmed fields message**

    To test:
    - Upload a document with a new field that is not yet in the schema
    - Navigate to the Relevance Tuning scree and you should see this message:

<img width="1705" alt="unconfirmed" src="https://user-images.githubusercontent.com/1427475/109033096-49c7cb80-7694-11eb-9c73-5f4ee64982b5.png">

6. **Added a schema conflicts message**

Note that I moved the position of this callout to be located with the other callouts, and removed the "dismiss" button, which seemed like overkill.

    To test:
    - Create a meta engine with two source engines that share a field with the same name, but each engine has the field in their schema as a different type
    - Navigate to this screen and you should see both a message at the top AND the fields listed at the bottom of the form

<img width="1709" alt="schema-conflicts" src="https://user-images.githubusercontent.com/1427475/109049937-d7acb200-76a6-11eb-8de7-b0197ab49e87.png">

7. **Add an empty state**

    To test:
    - Create a new engine with no documents
    - Navigate to this page and you should see the following message:

<img width="1711" alt="empty" src="https://user-images.githubusercontent.com/1427475/109055556-560c5280-76ad-11eb-8fa3-82795f7240d9.png">

8. **Un-title cases text in a few places, which was feedback from my previous PR**

### Checklist

Delete any items that are not applicable to this PR.


### For maintainers
